### PR TITLE
Enable String, URL and XML-related types for WASI

### PR DESCRIPTION
**Pull request by MaxDesiatov**
APIs related to file access (i.e. those depending on `NSCoding`) and multi-threading (those that require `NSRunLoop`) had to be disabled for WASI/Wasm to make this work.

### CODE REVIEW

**Comment by MaxDesiatov** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
Quick q: NSCoding only has a handful of places where it does file access; what's up with that?

**Comment by MaxDesiatov** [comment]
AFAIR NSCoding was tightly coupled with FS access, so it was easier to disable it. We also weren't expecting people to use APIs from Objective-C era on Wasm anyway. I can try and partially re-enable NSCoding w/o FS access if you think that makes more sense.

**Comment by MaxDesiatov** [comment]
@swift-ci please test

**Comment by MaxDesiatov** [comment]
@millenomi I've excluded changes related to `NSCoding` from this PR, ready for review

File: Sources/Foundation/URL.swift [hunk 1]
```diff
@@ -534,6 +535,11 @@ public struct URL : ReferenceConvertible, Equatable {
     public init(fileURLWithPath path: String) {
         _url = URL._converted(from: NSURL(fileURLWithPath: path.isEmpty ? "." : path))
     }
+#else
+    public init(fileURLWithPath path: String) {
+        _url = NSURL(string: "file://\(path)")!
```
[the remark this answers is not in the document]
Comment (karwa) [reply Sources/Foundation/URL.swift:540]: Also, this should at least check that path begins with a slash. Otherwise passing in "foo" results in the URL file://foo (i.e. with "foo" in the hostname position and no path)

File: Sources/FoundationXML/XMLParser.swift [hunk 2]
```diff
@@ -575,9 +588,33 @@ open class XMLParser : NSObject {
...
+    internal func parse(from data: Data) -> Bool {
+        var data = data
+        var result = true
+        let buffer = malloc(_chunkSize)!.bindMemory(to: UInt8.self, capacity: _chunkSize)
+        defer { free(buffer) }
+        var range = NSRange(location: 0, length: min(_chunkSize, data.count))
+        while result {
+            let chunk = data.withUnsafeMutableBytes { (rawBuffer: UnsafeMutableRawBufferPointer) -> Data in
+                let ptr = rawBuffer.baseAddress!.advanced(by: range.location)
+                return Data(bytesNoCopy: ptr, count: range.length, deallocator: .none)
```
Comment (karwa) [inline Sources/FoundationXML/XMLParser.swift:602]: This escapes a scoped inner-pointer. Since Data is self-slicing, can't you just use a regular slice?

File: Sources/FoundationXML/XMLParser.swift [hunk 3]
```diff
@@ -575,9 +588,33 @@ open class XMLParser : NSObject {
 
         return result
     }
-    
+#endif
+
+    internal func parse(from data: Data) -> Bool {
+        var data = data
+        var result = true
+        let buffer = malloc(_chunkSize)!.bindMemory(to: UInt8.self, capacity: _chunkSize)
```
Comment (karwa) [inline Sources/FoundationXML/XMLParser.swift:596]: Where is this used?

Comment (MaxDesiatov) [reply Sources/FoundationXML/XMLParser.swift:596]: It's used in `open func parse()` below.

Comment (karwa) [reply Sources/FoundationXML/XMLParser.swift:596]: I mean the buffer allocated on this line. It doesn’t seem to be used AFAICT.

**Comment by MaxDesiatov** [comment]
@swift-ci please test

**Comment by MaxDesiatov** [comment]
@millenomi I've cleaned it up, re-enabled as much as possible based on your feedback. There's still API surface relying on `InputStream`, I kept that unavailable. Do you think that's fine, or is keeping `InputStream` available on WASI a requirement to get this merged?

File: Sources/FoundationXML/XMLParser.swift [hunk 4]
```diff
@@ -993,8 +1031,10 @@ extension NSObject {
 func setupXMLParsing() {
     _CFSetupXMLInterface()
     _CFSetupXMLBridgeIfNeededUsingBlock {
+#if !os(WASI)
         __CFSwiftXMLParserBridge.CFBridge = CF.originalBridge
```
[the remark this answers is not in the document]
Comment (MaxDesiatov) [reply Sources/FoundationXML/XMLParser.swift]: I've re-enabled this in the updated commit

**Comment by MaxDesiatov** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
I feel the usage on Stream subclasses is low enough that I'm less worried.

File: Sources/Foundation/NSTextCheckingResult.swift [hunk 5]
```diff
@@ -53,7 +53,7 @@ open class NSTextCheckingResult: NSObject, NSCopying, NSSecureCoding {
     open class var supportsSecureCoding: Bool {
         NSRequiresConcreteImplementation()
     }
-    
+
```
Comment (xwu) [inline Sources/Foundation/NSTextCheckingResult.swift:56]: Is it possible to revert the unrelated whitespace changes?

File: Sources/Foundation/NSURL.swift [hunk 6]
```diff
@@ -841,6 +850,9 @@ extension NSURL {
     
     open func appendingPathComponent(_ pathComponent: String) -> URL? {
         var result : URL? = appendingPathComponent(pathComponent, isDirectory: false)
+
+        // File URLs can't be handled on WASI without file system access
+        #if !os(WASI)
```
Comment (xwu) [inline Sources/Foundation/NSURL.swift:855]:
```suggestion
#if !os(WASI)
```

File: Sources/Foundation/NSURL.swift [hunk 7]
```diff
@@ -852,6 +864,7 @@ extension NSURL {
             }
     
         }
+        #endif
```
Comment (xwu) [inline Sources/Foundation/NSURL.swift:867]:
```suggestion
#endif
```

File: Sources/Foundation/PropertyListSerialization.swift [hunk 8]
```diff
@@ -74,6 +74,7 @@ open class PropertyListSerialization : NSObject {
         }
     }
     
+    #if !os(WASI)
```
Comment (xwu) [inline Sources/Foundation/PropertyListSerialization.swift:77]: Similarly outdent here for consistency?

File: Sources/FoundationXML/XMLParser.swift [hunk 9]
```diff
@@ -575,9 +596,29 @@ open class XMLParser : NSObject {
...
+            chunkStart = chunkEnd
+            chunkEnd = min(chunkEnd + _chunkSize, data.count)
+        }
+        return result
+    }
+#endif
+
     // called to start the event-driven parse. Returns YES in the event of a successful parse, and NO in case of error.
     open func parse() -> Bool {
+        #if os(WASI)
```
Comment (xwu) [inline Sources/FoundationXML/XMLParser.swift:619]: Same comment here.

**Review by xwu — COMMENTED** [review]
Just some nits :)

File: Sources/Foundation/NSString.swift [hunk 10]
```diff
@@ -237,7 +241,7 @@ open class NSString : NSObject, NSCopying, NSMutableCopying, NSSecureCoding, NSC
     internal init(_ string: String) {
         _storage = string
     }
-    
+
```
Comment (MaxDesiatov) [inline Sources/Foundation/NSString.swift:244]:
```suggestion
    
```

**Comment by MaxDesiatov** [comment]
@swift-ci please test

**Comment by millenomi** [comment]
@MaxDesiatov I'm looking today.

**Comment by MaxDesiatov** [comment]
Thanks! I think I've addressed all of the current feedback, happy to update or answer any questions if anything else comes up.

**Comment by millenomi** [comment]
LGTM.
